spec/spec.go

```diff
@@ -2,81 +2,111 @@ package spec
+	Schemas map[string]*Schema `json:"schemas"`
+}
+
+type ExpansionResources struct {
+	OneOf []*Schema `json:"oneOf" yaml:"oneOf"`
+}
+
 type Fixtures struct {
 	Resources map[ResourceID]interface{} `json:"resources"`
 }
 
 type HTTPVerb string
 
-type JSONSchema struct {
-	Enum       []string               `json:"enum" yaml:"enum"`
-	Items      *JSONSchema            `json:"items" yaml:"items"`
-	OneOf      []*JSONSchema          `json:"oneOf" yaml:"oneOf"`
-	Properties map[string]*JSONSchema `json:"properties" yaml:"properties"`
-	Type       []string               `json:"type" yaml:"type"`
+type Schema struct {
+	AnyOf      []*Schema          `json:"anyOf" yaml:"anyOf"`
+	Enum       []string           `json:"enum" yaml:"enum"`
+	Items      *Schema            `json:"items" yaml:"items"`
+	Nullable   bool               `json:"nullable" yaml:"nullable"`
+	Properties map[string]*Schema `json:"properties" yaml:"properties"`
+	Required   []string           `json:"required" yaml:"required"`
+	Type       string             `json:"type" yaml:"type"`
 
 	// Ref is populated if this JSON Schema is actually a JSON reference, and
 	// it defines the location of the actual schema definition.
 	Ref string `json:"$ref" yaml:"$ref"`
 
-	XExpandableFields   []string    `json:"x-expandableFields" yaml:"x-expandableFields"`
-	XExpansionResources *JSONSchema `json:"x-expansionResources" yaml:"x-expansionResources"`
-	XResourceID         string      `json:"x-resourceId" yaml:"x-resourceId"`
-
-	// RawFields stores a raw map of JSON schema fields to values. This is
-	// useful when trying to interoperate with other libraries that use JSON
-	// schemas.
-	RawFields map[string]interface{}
+	XExpandableFields   []string            `json:"x-expandableFields" yaml:"x-expandableFields"`
+	XExpansionResources *ExpansionResources `json:"x-expansionResources" yaml:"x-expansionResources"`
+	XResourceID         string              `json:"x-resourceId" yaml:"x-resourceId"`
 }
 
-func (s *JSONSchema) UnmarshalJSON(data []byte) error {
-	type jsonSchema JSONSchema
-	var inner jsonSchema
-	err := json.Unmarshal(data, &inner)
+func (s *Schema) UnmarshalJSON(data []byte) error {
+	var rawFields map[string]interface{}
+	err := json.Unmarshal(data, &rawFields)
 	if err != nil {
 		return err
 	}
-	*s = JSONSchema(inner)
 
-	var rawFields map[string]interface{}
-	err = json.Unmarshal(data, &rawFields)
+	// If you update this, be sure to also update getJSONSchemaForOpenAPI3Schema
+	// to pass the fields through to the JSON validator
+	supportedFields := []string{
+		"anyOf", "enum", "items", "nullable", "properties", "required", "type",
+		"x-expandableFields", "x-expansionResources", "x-resourceId", "$ref",
+		"title", "description",
+	}
+	for _, supportedField := range supportedFields {
+		delete(rawFields, supportedField)
+	}
+	for unsupportedField := range rawFields {
+		return fmt.Errorf(
+			"unsupported field in JSON schema: '%s'", unsupportedField)
+	}
+
+	type schema2 Schema
```

Comment:
Mind if we find a slightly better name for this? Maybe `schemaAlias`?

Also, I know the pattern is pretty common in stripe-go, but it's still some pretty unintuitive sleight of hand. Mind adding a comment that explains that this is here to avoid a recursive call into this `UnmarshalJSON` function?